internal/server/grants_test.go

```diff
@@ -344,6 +345,178 @@ func TestAPI_ListGrants(t *testing.T) {
 	}
 }
 
+func TestAPI_ListGrants_InheritedGrants(t *testing.T) {
+	srv := setupServer(t, withAdminUser)
+	routes := srv.GenerateRoutes(prometheus.NewRegistry())
+
+	createID := func(t *testing.T, name string) uid.ID {
+		t.Helper()
+		var buf bytes.Buffer
+		body := api.CreateUserRequest{Name: name}
+		err := json.NewEncoder(&buf).Encode(body)
+		assert.NilError(t, err)
+
+		req, err := http.NewRequest(http.MethodPost, "/api/users", &buf)
+		assert.NilError(t, err)
+		req.Header.Add("Authorization", "Bearer "+adminAccessKey(srv))
+		req.Header.Add("Infra-Version", "0.12.3")
+
+		resp := httptest.NewRecorder()
+		routes.ServeHTTP(resp, req)
+		assert.Equal(t, resp.Code, http.StatusCreated, resp.Body.String())
+		respObj := &api.CreateUserResponse{}
+		err = json.Unmarshal(resp.Body.Bytes(), respObj)
+		assert.NilError(t, err)
+		return respObj.ID
+	}
+
+	createGroup := func(t *testing.T, name string, users ...uid.ID) uid.ID {
+		t.Helper()
+		group := &models.Group{Name: name}
+
+		err := data.CreateGroup(srv.db, group)
+		assert.NilError(t, err)
+
+		err = data.AddUsersToGroup(srv.db, group.ID, users)
+		assert.NilError(t, err)
+
+		return group.ID
+	}
+
+	idInGroup := createID(t, "inagroup@example.com")
+	mikhail := createID(t, "mikhail@example.com")
+
+	zoologistsID := createGroup(t, "Zoologists", mikhail)
+
+	var accessKey string
+
+	loginAs := func(userID uid.ID) {
+		token := &models.AccessKey{
+			IssuedFor:  userID,
+			ProviderID: data.InfraProvider(srv.db).ID,
+			ExpiresAt:  time.Now().Add(10 * time.Minute),
+		}
+
+		var err error
+		accessKey, err = data.CreateAccessKey(srv.db, token)
```

Comment:
This modification of `accessKey` is really surprising, and is going to make test failures hard to debug. The next test case added to this test function will use this `accessKey` instead of the default one.

There should be an easy way to avoid this by changing this to modify the `http.Request`:


```go
func loginAs(db *gorm.DB, userID uid.ID) func(t *testing.T, req *http.Request) {
    ... // the current logic in this function
    return func(t *testing.T, req *http.Request) {
        req.Header.Set("Authorization", "Bearer "+accessKey)
    }
}
```

This would allow `loginAs` to be used directly as the setup function.  But there are lots of other options that allow you to create an access key directly inline and pass it to the request, without  modifying variables that are used in other test cases.

Reply:
nice suggestion.